feat(formatter): Detect code removal feature

leaysgur opened the pull request:
Part of #13720 

- Introduce `detect_code_removal` feature flag
- If enabled, `Formatter::format()` performs extra check
  - Currently, every public API calls `format()`
- Code removal detection consists of the following two steps:
  - `bf = collect(original_text)` and `af = collect(formatted_text)`
  - `diff(bf, af)`
- Basically, this works by counting `node.debug_name()`
  - Some AST nodes are ignored due to difficulty

NOTE: The feature has been added but is not yet enabled anywhere.

---

- Enable feature somewhere
  - tasks/prettier_conformance
  - tasks/coverage
  - tasks/benchmark
  - crates/oxc_formatter
  - crates/oxc_language_server
  - apps/oxfmt
  - napi/playground
  - monitor-oxc
  - oxlint-ecosystem-ci
- This feature should be enabled by default and should be opt-out, or explicitly enabled?

leaysgur commented:
* **#15059** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/15059?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/15059?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=15059).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Dunqing commented:
Leaves `oxlint-ecosystem-ci` to me, I am making some changes to it.

leaysgur commented on crates/oxc_formatter/src/lib.rs:
```diff
@@ -87,6 +89,25 @@ impl<'a> Formatter<'a> {
+
+            #[expect(clippy::print_stderr)]
+            if let Some(diff) = detect_code_removal::diff(&before_stats, &after_stats) {
+                eprintln!("🚨 Code removal detected during formatting:\n{diff}");
```
Updated.

Dunqing commented:
### Merge activity

* **Oct 31, 12:47 AM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.
* **Oct 31, 12:47 AM UTC**: `Dunqing` added this pull request to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc).

leaysgur committed:
feat(formatter): Detect code removal feature (#15059)

Part of #13720

- Introduce `detect_code_removal` feature flag
- If enabled, `Formatter::format()` performs extra check
  - Currently, every public API calls `format()`
- Code removal detection consists of the following two steps:
  - `bf = collect(original_text)` and `af = collect(formatted_text)`
  - `diff(bf, af)`
- Basically, this works by counting `node.debug_name()`
  - Some AST nodes are ignored due to difficulty

NOTE: The feature has been added but is not yet enabled anywhere.

---

- Enable feature somewhere
  - tasks/prettier_conformance
  - tasks/coverage
  - tasks/benchmark
  - crates/oxc_formatter
  - crates/oxc_language_server
  - apps/oxfmt
  - napi/playground
  - monitor-oxc
  - oxlint-ecosystem-ci
- This feature should be enabled by default and should be opt-out, or explicitly enabled?
